src/contracts/core/StakeRootCompendium.sol

```diff
@@ -26,170 +25,139 @@ contract StakeRootCompendium is StakeRootCompendiumStorage {
+        _removeOperatorSetFromStakeTree(IAVSDirectory.OperatorSet({avs: msg.sender, operatorSetId: operatorSetId}));
     }
 
     /// @inheritdoc IStakeRootCompendium
     function setOperatorSetExtraData(uint32 operatorSetId, bytes32 extraData) external {
-        (bool exists,, uint224 index) = operatorSetToIndex[msg.sender][operatorSetId].latestCheckpoint();
-        require(exists && index != REMOVED_INDEX, "StakeRootCompendium.setOperatorSetExtraData: operatorSet is not in stakeTree");
         operatorSetExtraDatas[msg.sender][operatorSetId] = extraData;
     }
 
     /// @inheritdoc IStakeRootCompendium
     function setOperatorExtraData(uint32 operatorSetId, address operator, bytes32 extraData) external {
-        (bool exists,, uint224 index) = operatorSetToIndex[msg.sender][operatorSetId].latestCheckpoint();
-        require(exists && index != REMOVED_INDEX, "StakeRootCompendium.setOperatorExtraData: operatorSet is not in stakeTree");
         operatorExtraDatas[msg.sender][operatorSetId][operator] = extraData;
     }
 
     /// POSTING ROOTS AND BLACKLISTING
 
     /// @inheritdoc IStakeRootCompendium
     function verifyStakeRoot(
-        uint32 calculationTimestamp,
+        uint64 calculationBlockNumber,
         bytes32 stakeRoot,
         address chargeRecipient,
         Proof calldata proof
     ) external {
-        // TODO: verify proof
-
-        _postStakeRoot(calculationTimestamp, stakeRoot, chargeRecipient, false);
-    }
-
-    /// @inheritdoc IStakeRootCompendium
-    function blacklistStakeRoot(uint32 submissionIndex) external onlyOwner {
-        // TODO: this should not be onlyOwner
         require(
-            !stakeRootSubmissions[submissionIndex].blacklisted,
-            "StakeRootCompendium.blacklistStakeRoot: stakeRoot already blacklisted"
+            calculationBlockNumber % proofInterval == 0,
+            "StakeRootCompendium._postStakeRoot: blockNumber must be a multiple of proofInterval"
         );
         require(
-            block.timestamp < stakeRootSubmissions[submissionIndex].blacklistableBefore,
-            "StakeRootCompendium.blacklistStakeRoot: stakeRoot cannot be blacklisted"
+            latestBlockNumberProven < calculationBlockNumber,
+            "StakeRootCompendium._postStakeRoot: calculationBlockNumber must be greater than the last posted calculationBlockNumber"
         );
         require(
-            !stakeRootSubmissions[submissionIndex].forcePosted,
-            "StakeRootCompendium.blacklistStakeRoot: stakeRoot was force posted"
+            stakeRootSubmissions[calculationBlockNumber].proven == false, 
+            "StakeRootCompendium._postStakeRoot: calculationBlockNumber already proven"
         );
-        stakeRootSubmissions[submissionIndex].blacklisted = true;
+        // TODO: verify proof
+        _postStakeRoot(true, calculationBlockNumber, stakeRoot);
+        uint256 charge = totalChargeSnapshot.upperLookup(uint32(calculationBlockNumber));
```

Comment:
should we just allow them to submit an index here to avoid lookup gas?

Reply:
i need to put a bit more thought into the total charge, thats a component I overlooked a bit. . .

can't we avoid snapshots by having a view function for the totalCharge and doing a storage proof for that block? that's basically what the zk proof is doing anyways right? could add that to the proof? and to avoid zk tech stack failure enforce a MAX_TOTAL_CHARGE = 3 eth

alternatively to KISS we can do a little bit more accounting to have an O(1) lookup, I think there's a simple way to do it